exploration/function-composition-part-1.md

```diff
@@ -0,0 +1,974 @@
+a formatted number, which `:number` does not accept.
+
+The answer to this question is unclear.
+How different values for the same options might
+(or might not) combine is not straightforward.
+Perhaps some do, others don't, and the details are
+specific to each function.
+
+Example A6:
+Should $y be formatted as ٩٨٧٦٥ or 98765?
+
+```
+    input $num = 98765
+    .local $x = {$num :number numberSystem=arab}
+    .local $y = {$x :number numberSystem=latn}
+```
+
+Example A7:
+Should $y be formatted as +98765 or 98765?
+```
+
+
+    input $num = 98765
+    .local $x = {$num :number signDisplay=always}
+    .local $y = {$x :number signDisplay=auto}
+```
+
+These examples raise the question
+of whether functions return values
+that encode the option names and values
+that were passed in to the function.
+
+While some use cases don't work well
+(or at least work surprisingly)
+if options are **not** encoded in named values,
+defining what it means to compose options
+is not straightforward.
+
+### Combining options
+
+In Example Y1,
+two function calls are composed
+with the same set of option names
+(and different option values).
+
+A question also arises of how to combine options with different names.
+
+Example A4:
+
+```
+    .local $x = {|1| :number minInt=3}
+    .local $y = {$x :number maxFrac=5}
+    {{$x} {$y}}
+```
+
+Is this message equivalent to Example A5?
+
+Example A5:
+
+```
+    .local $x = {|1| :number maxFrac=5 minInt=3}
+    {{$x} {$x}}
+```
+
+In compositions of calls to the same function,
+it might make intuitive sense to union together
+the option sets, letting the outermost enclosing
+call take precedence if the same option is specified
+multiple times.
+
+It's less obvious what to do with compositions of
+calls to different functions, which may have different
+option sets.
+
+
+### Computation vs. applying formatting
+
+The previous examples conceptually return
+the same value that was passed in,
+annotated with formatting hints.
+
+But functions can do other kinds of computation.
+
+Example B1:
+```
+    .local $age = {$person :getAge}
+    .local $y = {$age :duration skeleton=yM}
+    .local $z = {$y :uppercase}
+```
+
+Although there is also a pipeline of functions
+(conceptually, `uppercase . duration . getAge`),
+the formatted value returned by `:getAge` is _not_
+just "the argument with formatting options applied",
+but rather, a piece of the argument.
+Other functions can be imagined that do more general
+computation on arguments.
+
+It would not be correct to say that
+`:uppercase` converts a person to uppercase,
+nor would it be correct to say that `:uppercase`
+converts a number to uppercase.
+
+This example only makes sense if `:uppercase`
+operates on the "formatted result"
+of evaluating the _expression_ bound to `$y`.
+
+This suggests a representation for named values
+that allows functions to choose whether to
+inspect the "formatted result", the "argument" and options,
+or both.
```

Comment:
optional, although I think this would help clarity for when drafting a future design doc that takes a position, since we would be less likely to forget options that are on the table
```suggestion
or both.
Or, it may also suggest that we consider 
whether we allow composition at all when the functions differ.
```